src/WebHandlers.cpp

```diff
@@ -23,73 +23,86 @@
-  }
-  //  if the root of the request matches the _uri then it checks to see if there is a file it can handle. 
-  if (request->url().startsWith(_uri)) {
-    String path = _getPath(request);
-    if (_fs.exists(path) || _fs.exists(path + ".gz")) {
-       DEBUGF("[AsyncStaticWebHandler::canHandle] TRUE\n");
-      return true;
-    }
+  if (request->method() == HTTP_GET &&
+      request->url().startsWith(_uri) &&
+      _getPath(request, true).length()) {
+
+    DEBUGF("[AsyncStaticWebHandler::canHandle] TRUE\n");
+    return true;
+
   }
 
   return false;
 }
 
-String AsyncStaticWebHandler::_getPath(AsyncWebServerRequest *request)
+String AsyncStaticWebHandler::_getPath(AsyncWebServerRequest *request, const bool withStats)
 {
+  // Remove the found uri
+  String path = request->url().substring(_uri.length());
+
+  // We can skip the file check if we serving a directory and (we have full match or we end with '/')
+  bool canSkipFileCheck = _isDir && (path.length() == 0 || path[path.length()-1] == '/');
 
-  String path = request->url();
-  DEBUGF("[AsyncStaticWebHandler::_getPath]\n");
-  DEBUGF("  [stored] _uri = %s, _path = %s\n" , _uri.c_str(), _path.c_str() ) ;
-  DEBUGF("  [request] url = %s\n", request->url().c_str() );
+  path = _path + path;
 
-  if (!_isFile) {
-    DEBUGF("   _isFile = false\n");
-    String baserequestUrl = request->url().substring(_uri.length());  // this is the request - stored _uri...  /espman/
-    DEBUGF("  baserequestUrl = %s\n", baserequestUrl.c_str());
+  // Do we have a file or .gz file
+  if (!canSkipFileCheck) if (_fileExists(path, withStats)) return path;
 
-    if (!baserequestUrl.length()) {
-      baserequestUrl += "/";
-    }
+  // Try to add default page, ensure there is a trailing '/' ot the path.
+  if (path.length() == 0 || path[path.length()-1] != '/') path += "/";
+  path += "index.htm";
 
-    path = _path + baserequestUrl;
-    DEBUGF("  path = path + baserequestUrl, path = %s\n", path.c_str());
+  if (_fileExists(path, withStats)) return path;
+
+  // No file - return empty string
+  return String();
+}
+
+bool AsyncStaticWebHandler::_fileExists(const String path, const bool withStats)
+{
+  bool fileFound = false;
+  bool gzipFound = false;
 
-    if (path.endsWith("/")) {
-      DEBUGF("  3 path ends with / : path = index.htm \n");
-      path += "index.htm";
-    }
+  String gzip = path + ".gz";
+
+  if (_gzipFirst) {
+    gzipFound = _fs.exists(gzip);
+    if (!gzipFound) fileFound = _fs.exists(path);
   } else {
-    path = _path;
+    fileFound = _fs.exists(path);
+    if (!fileFound) gzipFound =  _fs.exists(gzip);
   }
 
-  DEBUGF(" final path = %s\n", path.c_str());
-  DEBUGF("[AsyncStaticWebHandler::_getPath] END\n\n");
+  bool found = fileFound || gzipFound;
+
+  if (withStats && found) {
+    _gzipStats = (_gzipStats << 1) + gzipFound ? 1 : 0;
+    _fileStats = (_fileStats << 1) + fileFound ? 1 : 0;
+    _gzipFirst = _countBits(_gzipStats) > _countBits(_fileStats);
+  }
 
```

Comment:
_gzipStats and _fileStats are shifted left every time we find a file, the LSB is set for _gzipStats if we have found a gzip file, or to _fileStats if we found a regular file, but never to both of them.
For example: after serving gzip, file, gzip, gzip, file, gzip. _gzipStats = 0b00101101 and _fileStats = 0b00010010. We then count the number of set bits in each and if we served more gzip files _gzipFirst is set so the code will do exists(path+'.gz') before exists(path) to avoid the extra call in most cases.
A faster option is to use just one byte _gzipStats and set _gzipFirst =  _countBits(_gzipStats) > 4, but this will require 4 to 8 files to warm up:
`
+  if (withStats && found) {
+    _gzipStats = (_gzipStats << 1) + gzipFound ? 1 : 0;
+    _gzipFirst = _countBits(_gzipStats) > 4;
+  }
`